### PR TITLE
[SPARK-12448][SQL] Add UserDefinedType support to Cast

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-12448

We should add UserDefinedType support to Cast.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 1]
```diff
@@ -790,4 +820,25 @@ class CastSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+  test("cast between UserDefinedTypes") {
+    assert(Cast.canCast(new ExampleUDT, new ExampleUDT) == true)
+    assert(Cast.canCast(new ExampleUDT, new Example2UDT) == false)
+
+    val udt = new ExampleClass
+
+    val castExpression = Cast(Literal.create(udt, new ExampleUDT), new ExampleUDT)
+    checkEvaluationWithoutCodegen(castExpression, udt)
+    val castExpression2 = Cast(Literal.create(0.0, new ExampleUDT), new ExampleUDT)
+    checkEvaluationWithGeneratedMutableProjection(castExpression2, 0.0)
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: For codgen case, the internal data for ExampleUDT here is double.

**Comment by marmbrus** [comment]
I think we should hold off on this one too until we have a proper design for UDTs.

**Comment by viirya** [comment]
@marmbrus This is related to #10390. Do I need to submit them together in #10390 or leave this as separated one? Which is one better?

**Comment by marmbrus** [comment]
We can just merge the other PR once comments are addressed.

**Comment by viirya** [comment]
We don't need this pr now. Close it.

**Pull request closed without merging**